Expand docs for lexical structure

mosabua opened the pull request:
## Description

Document lexical structure and add necessary updates to other pages. We should talk about how to proceed @martint since I believe the linked page for reserved keywords is also wrong by now and there are other things I have questions about how we want to document all this. 


<!-- Provide details that would help an engineer who is unfamiliar with this part of the code. -->
## Additional context and related issues

* Triggered by #17776
* Replaces #18000 

<!-- Mark the appropriate option with an (x). Propose a release note if you can. -->
## Release notes

(x) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

nineinchnick commented on docs/src/main/sphinx/language/lexical-structure.rst:
```diff
@@ -0,0 +1,135 @@
+Statements can also contain comments, but they are not tokens, so they are
+effectively ignored.
+
+The following example contains all elements::
```
Is the double colon intentional?

nineinchnick commented on docs/src/main/sphinx/language/lexical-structure.rst:
```diff
@@ -0,0 +1,135 @@
+
+The following example contains all elements::
+
+    SELECT * FROM example_table WHERE a_column = '5'; -- a comment
```
Use a quoted identifier to make the example complete
```suggestion
    SELECT "other column" FROM example_table WHERE a_column = '5'; -- a comment
```

nineinchnick commented on docs/src/main/sphinx/language/lexical-structure.rst:
```diff
@@ -0,0 +1,135 @@
+used for the supported :doc:`/sql`.
+
+Identifiers must start with a letter and subsequently include letters, digits,
+and underscores. A dot is used to compose multiple identifiers to a of a
```
```suggestion
and underscores. A dot is used to compose multiple identifiers to a
```

nineinchnick commented on docs/src/main/sphinx/language/lexical-structure.rst:
```diff
@@ -0,0 +1,135 @@
+
+Identifiers must start with a letter and subsequently include letters, digits,
+and underscores. A dot is used to compose multiple identifiers to a of a
+qualified identifier. Identifiers with special characters must be quoted.
```
People asked a few times on Slack what characters are allowed in quoted identifiers, so it's worth explaining that anything is valid and only double quotes need to be escaped.
```suggestion
qualified identifier. Identifiers with special characters must be quoted.
A quoted identifier can contain any characters. If it includes a double quote
character, it needs to be escaped with another double quote character.
```

nineinchnick commented on docs/src/main/sphinx/language/lexical-structure.rst:
```diff
@@ -0,0 +1,135 @@
+Identifiers must start with a letter and subsequently include letters, digits,
+and underscores. A dot is used to compose multiple identifiers to a of a
+qualified identifier. Identifiers with special characters must be quoted.
+Identifiers are not treated as case sensitive. Following are some valid
```
```suggestion
Identifiers and quoted identifiers are not treated as case sensitive. Following are some valid
```

nineinchnick commented on docs/src/main/sphinx/language/lexical-structure.rst:
```diff
@@ -0,0 +1,135 @@
+^^^^^^^^^^^^^^^
+
+Single quotes, escaping them.
+Unicode Escapes
```
We could start by copying this paragraph: https://trino.io/docs/current/language/types.html#varchar

Also, maybe it's worthing referencing this page in the data types section? A few times I was looking for those unicode escape codes in docs and couldn't remember where they are.

nineinchnick commented on docs/src/main/sphinx/language/types.md:
```diff
@@ -41,6 +41,22 @@ This type captures boolean values `true` and `false`.
+increase readability. For example, decimal integer `123_456.789_123` is
+equivalent to `123456.789123`. Preceding underscores are not permitted.
+
+Integers are supported by the following data types with increasing size range.
```
Wdyt about expressing min and max values for different integer types using hex notation? It's easier for me to remember that max is just a different number of f's, like 0x7fff for a small int, and min is different number of zeros, ex. -0x800

mosabua replied:
I think thats a good idea .. in addition to the current values 

From 

```
a minimum value of `-2^7` and a maximum value of `2^7 - 1`.
```

nineinchnick replied:
-0x8000 and 0x7FFF for small int, 2 digits for every byte, so 32 bit integer is 4 bytes and max for it would be 0x7FFFFFFF (8 characters after 0x). So much easier to remember than 2,147,483,647 😂

mosabua replied:
Added now .. hoping I got it right..

nineinchnick commented on docs/src/main/sphinx/language/reserved.md:
```diff
@@ -89,3 +92,32 @@ be quoted (using double quotes) in order to be used as an identifier.
+
+Identifiers must start with a letter, and subsequently include alphanumeric
+characters and underscores. Identifiers with other characters must be delimited
+with double quotes (`"`). When delimited with double quotes, identifiers can use
```
This should mention how to escape a double quote in a quoted identifier.

mosabua replied:
its `\"` right?

martint replied:
No, you escape a `"` with another double quote. I.e., `""`. This analogous to how single quotes are escaped in string literals.

nineinchnick replied:
It's double the double quote

martint commented on docs/src/main/sphinx/language/types.md:
```diff
@@ -41,6 +41,22 @@ This type captures boolean values `true` and `false`.
+* Binary integer composed of `0B` or `0b` and the value. Examples are `0b1001`
+  for decimal `9` or `0b101010` for decimal `42``.
+
+Underscore characters are ignored within literal values, and can be used to
```
However, underscores are only allowed in certain places within a literal.

For example, these are invalid numeric literals:

`1_`
`_1`
`1_.234`
`1._234`
`1.234_`
`0x1_`
`0_x1`

martint commented on docs/src/main/sphinx/language/types.md:
```diff
@@ -66,6 +82,14 @@ A 64-bit signed two's complement integer with a minimum value of
 
+Variable-precision numbers can be expressed as numeric literal using the
+scientific notation such as `1.03e1` for the `DOUBLE` data type. The supported
+`REAL` and `DOUBLE` data types also support explicit declaration of the type.
```
What does this mean?

mosabua replied:
Spelled it out more now.

mosabua commented:
Addressed all feedback now. 

@nineinchnick can you check the values and also confirm this is still useful despite there being LOTS of digits now.

@martint anything else I missed we should add or other feedback?

mosabua commented on docs/src/main/sphinx/language/reserved.md:
```diff
@@ -89,3 +92,34 @@ be quoted (using double quotes) in order to be used as an identifier.
+Identifiers must start with a letter, and subsequently include alphanumeric
+characters and underscores. Identifiers with other characters must be delimited
+with double quotes (`"`). When delimited with double quotes, identifiers can use
+any character. Escape a `"` with another preceding double quote in in a
```
hahah .. fixed

mosabua commented:
Any other requests @martint ?

martint commented on docs/src/main/sphinx/language/types.md:
```diff
@@ -41,31 +41,58 @@ This type captures boolean values `true` and `false`.
+Underscore characters are ignored within literal values, and can be used to
+increase readability. For example, decimal integer `123_456.789_123` is
+equivalent to `123456.789123`. Preceding underscores are not permitted.
+Preceding underscores and trailing underscores are not permitted.
```
This is partially redundant. It repeats the previous sentence.

mosabua replied:
Oops .. updated

martint commented on docs/src/main/sphinx/language/types.md:
```diff
@@ -41,31 +41,58 @@ This type captures boolean values `true` and `false`.
+equivalent to `123456.789123`. Preceding underscores are not permitted.
+Preceding underscores and trailing underscores are not permitted.
+
+Integers are supported by the following data types with increasing size range.
```
What's "size range"? I would just remove the "with increasing size range clause". It doesn't add much and the language is confusing.

mosabua replied:
Removed

martint commented on docs/src/main/sphinx/language/types.md:
```diff
@@ -41,31 +41,58 @@ This type captures boolean values `true` and `false`.
 
 ## Floating-point
 
+Variable-precision numbers can be expressed as numeric literal using the
```
"using scientific notation" (remove *the*)

mosabua replied:
Done

martint commented on docs/src/main/sphinx/language/types.md:
```diff
@@ -41,31 +41,58 @@ This type captures boolean values `true` and `false`.
 
+Variable-precision numbers can be expressed as numeric literal using the
+scientific notation such as `1.03e1` for the `DOUBLE` data type. The supported
+`REAL` and `DOUBLE` data types also support explicit declaration of the type
```
> also support explicit declaration of the type with the keywords `REAL` or `DOUBLE` and a value.

What does this mean? If this is referring to the `DOUBLE '1.23e4'` syntax. There's nothing special about these types -- that's just the generic type constructor syntax, and it applies to every type unless the type has no textual representation.

mosabua replied:
well... if you dont have the type defined .. you get double automatically .. how do we say that  .. or do we just leave it as implied from the following sections?

martint replied:
Those are two different things. Numeric scientific precision *literals* have a type of `DOUBLE`. E.g.,

```
1.23e4
```

Values of `DOUBLE` and `REAL` (and most other types) can be constructed via their type constructor:

```
DOUBLE '123e4'
REAL '123e4'
VARCHAR 'abcde'
DECIMAL '123.4'
TIMESTAMP '2023-08-21 12:34:56'
etc.
```

mosabua replied:
I think I updated this now so it makes more sense and is clear.

martint commented on docs/src/main/sphinx/language/types.md:
```diff
@@ -41,31 +41,58 @@ This type captures boolean values `true` and `false`.
+with the keywords `REAL` or `DOUBLE` and a value.
+
+Underscore characters are ignored within literal values, and can be used to
+increase readability. For example, variable-precision `123_456.789e4` is
```
> For example, *the value* `123_456.789e4`

mosabua replied:
Updated

mosabua commented:
All updated apart from the question about REAL and DOUBLE

mosabua commented:
Ping @martint

martint commented on docs/src/main/sphinx/language/types.md:
```diff
@@ -41,31 +41,61 @@ This type captures boolean values `true` and `false`.
 
 ## Floating-point
 
+Floating-point, or variable-precision, numbers can be expressed as numeric
```
"variable precision" is incorrect. REAL and DOUBLE have fixed precision.

mosabua replied:
updated

martint commented on docs/src/main/sphinx/language/types.md:
```diff
@@ -41,31 +41,61 @@ This type captures boolean values `true` and `false`.
+`123456.789e4`. Preceding underscores, trailing underscores, and underscores
+beside the comma (`.`) are not permitted.
+
+The supported `REAL` and `DOUBLE` data types also support explicit declaration
+of the type with the keywords `REAL` or `DOUBLE` and a value.
```
The grammar here is off. I would just get rid of this and the next paragraph, as I suggested in a previous discussion. What this describes is a property of pretty much every type. If we want to talk about it, would add a section somewhere in the docs that explains type constructor syntax.

mosabua commented:
Ping @martint
